tests/Unit/Evolution/Systems/GeneralizedHarmonic/Test_Characteristics.cpp

```diff
@@ -0,0 +1,546 @@
+}
+
+// Test return-by-reference GH char speeds by comparing to Kerr-Schild
+template <typename Solution>
+void test_characteristic_speeds_analytic(
+    const Solution& solution, const size_t grid_size_each_dimension,
+    const std::array<double, 3>& lower_bound,
+    const std::array<double, 3>& upper_bound) noexcept {
+  // Set up grid
+  const size_t spatial_dim = 3;
+  const size_t data_size = pow<spatial_dim>(grid_size_each_dimension);
+  Mesh<spatial_dim> mesh{grid_size_each_dimension, Spectral::Basis::Legendre,
+               Spectral::Quadrature::GaussLobatto};
+
+  using Affine = CoordinateMaps::Affine;
+  using Affine3D = CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1., 1., lower_bound[0], upper_bound[0]},
+          Affine{-1., 1., lower_bound[1], upper_bound[1]},
+          Affine{-1., 1., lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-independent solution.
+  const double t = std::numeric_limits<double>::signaling_NaN();
+
+  // Evaluate analytic solution
+  const auto vars =
+      solution.variables(x, t, typename Solution::template tags<DataVector>{});
+  const auto& lapse = get<gr::Tags::Lapse<>>(vars);
+  const auto& shift = get<gr::Tags::Shift<spatial_dim>>(vars);
+  const auto& spatial_metric = get<gr::Tags::SpatialMetric<spatial_dim>>(vars);
+
+  // Get ingredients
+  const auto& gamma_1 = make_with_value<Scalar<DataVector>>(x, 0.1);
+  const auto& inverse_spatial_metric =
+      determinant_and_inverse(spatial_metric).second;
+  // Outward 3-normal to the surface on which characteristic fields are needed
+  auto unit_normal_one_form =
+      make_with_value<tnsr::i<DataVector, spatial_dim, Frame::Inertial>>(x, 1.);
+  const auto norm_of_one_form = sqrt(get(dot_product(unit_normal_one_form,
+      unit_normal_one_form, inverse_spatial_metric)));
+  for (size_t i = 0; i < spatial_dim; ++i) {
+    unit_normal_one_form.get(i) /= norm_of_one_form;
+  }
+
+  // Get generalized harmonic characteristic speeds locally
+  const auto shift_dot_normal = get(dot_product(shift, unit_normal_one_form));
+  const auto upsi_speed = -(1. + get(gamma_1)) * shift_dot_normal;
+  const auto uzero_speed = -shift_dot_normal;
+  const auto uplus_speed = -shift_dot_normal + get(lapse);
+  const auto uminus_speed = -shift_dot_normal - get(lapse);
+
+  // Check that locally computed fields match returned ones
+  const auto& char_speeds_from_func =
+      GeneralizedHarmonic::CharacteristicSpeedsCompute<spatial_dim,
+          Frame::Inertial>::function(gamma_1, lapse, shift,
+              unit_normal_one_form);
+  const auto& upsi_speed_from_func =
+      get(get<Tags::CharSpeed<GeneralizedHarmonic::Tags::UPsi<spatial_dim,
```

Comment:
No big deal, but can you simplify this using your `compute_speed_with_tag` function above?

Reply:
That would require four computations of all characteristic speeds, following by only one of them being extracted and used each time. Right now, all fields are computed once, but need to be extracted from their container before use. Is a reduction of `5` to `4` function calls worth quadrupling the computation of this block?